### PR TITLE
Fix small bugs and add solutions

### PR DESCRIPTION
**Pull request by fdw**
* rename already defined methods (by adding a ')
* import a module that is necessary for readMaybe - I always thought I was doing something wrong, until Google helped me ;)
* add solutions. Probably not the most elegant ones, but at least they work. Feedback is welcome!

### CODE REVIEW

File: beginner-workshop_rolling-update/src/exercises/Solutions.hs [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+-- Prepend (»cons«) two elements to the front of a list. Hint: (:) does it for
+-- a single element,
+--
+-- >>> 1 : [2,3]
+-- [1,2,3]
+--
+-- >>> cons2 1 2 [3,4,5]
+-- [1,2,3,4,5]
+cons2 x y list = x : (y : list)
```
Comment (quchen) [inline beginner-workshop_rolling-update/src/exercises/Solutions.hs]: Redundant parenthesis

File: beginner-workshop_rolling-update/src/exercises/Solutions.hs [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+-- Get the second element of a list, if there is one.
+--
+-- >>> head2 [1..10]
+-- Just 2
+-- >>> head2 [1]
+-- Nothing
+head2 :: [a] -> Maybe a
+head2 [] = Nothing
+head2 (x:y:xs) = Just y
```
Comment (quchen) [inline beginner-workshop_rolling-update/src/exercises/Solutions.hs]: x and xs are unused, use a wildcard pattern: (_:x:_)

File: beginner-workshop_rolling-update/src/exercises/Solutions.hs [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+
+-- Many solutions are possible, such as
+--   - Using guards
+--   - Pattern matching
+--   - Using the built-in `product`
+factorial :: Integer -> Integer
+factorial x
+  | x == 0 = 0
+  | x == 1 = 1
+  | otherwise = x * factorial (x-1)
```
Comment (quchen) [inline beginner-workshop_rolling-update/src/exercises/Solutions.hs]: What happens for x<0? The function claims to work for all integers, after all.

File: beginner-workshop_rolling-update/src/exercises/Solutions.hs [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+mergesort :: Ord a => [a] -> [a]
+mergesort xs
+  | length xs == 0 = []
+  | length xs == 1 = xs
+  | otherwise = let (ys, zs) = splitIntoHalves xs
+    in merge (mergesort ys) (mergesort zs)
+  where
+    -- Merge two ordered lists into one
+    merge :: Ord a => [a] -> [a] -> [a]
+    merge (x:xs) (y:ys) = if x < y then x:(merge xs (y:ys)) else y:(merge (x:xs) ys)
```
Comment (quchen) [inline beginner-workshop_rolling-update/src/exercises/Solutions.hs]: Redundant parentheses

File: beginner-workshop_rolling-update/src/exercises/Solutions.hs [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+find' predicate (y:ys) = if predicate y then Just y else find' predicate ys
+find' predicate [] = Nothing
+
+
+-- Pair entries up, stop when one list is empty
+-- e.g. zip' [1..] "abc" ==> [(1,'a'), (2,'b'), (3,'c')]
+zip' :: [a] -> [b] -> [(a,b)]
+zip' (x:xs) (y:ys) = [(x, y)] ++ (zip' xs ys)
+zip' [] _ = []
+zip' _ [] = []
```
Comment (quchen) [inline beginner-workshop_rolling-update/src/exercises/Solutions.hs]: You can combine the last two cases

File: beginner-workshop_rolling-update/src/exercises/Solutions.hs [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+-- Get the first element matching the predicate
+find' :: (a -> Bool) -> [a] -> Maybe a
+find' predicate (y:ys) = if predicate y then Just y else find' predicate ys
+find' predicate [] = Nothing
+
+
+-- Pair entries up, stop when one list is empty
+-- e.g. zip' [1..] "abc" ==> [(1,'a'), (2,'b'), (3,'c')]
+zip' :: [a] -> [b] -> [(a,b)]
+zip' (x:xs) (y:ys) = [(x, y)] ++ (zip' xs ys)
```
Comment (quchen) [inline beginner-workshop_rolling-update/src/exercises/Solutions.hs]: [x] ++ = x :

File: beginner-workshop_rolling-update/src/exercises/Solutions.hs [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+--
+-- >>> zipWith (+) [1,2,3] [10,20,30]
+-- [11,22,33]
+--
+-- NB: zip = zipWith (,)
+--     zipWith' f = map f . zip
+zipWith' :: (a -> b -> c) -> [a] -> [b] -> [c]
+zipWith' f (x:xs) (y:ys) = [(f x y)] ++ zipWith f xs ys
+zipWith' _ [] _ = []
+zipWith' _ _ [] = []
```
Comment (quchen) [inline beginner-workshop_rolling-update/src/exercises/Solutions.hs]: Dito ×2

File: beginner-workshop_rolling-update/src/exercises/Solutions.hs [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+--
+-- >>> runAll [putStrLn "hello", putStrLn "world"]
+-- hello
+-- world
+runAll :: [IO a] -> IO ()
+runAll [] = return ()
+runAll (x:xs) = do 
+  _ <- x
+  runAll xs
+  return ()
```
Comment (quchen) [inline beginner-workshop_rolling-update/src/exercises/Solutions.hs]: This line is redundant

File: beginner-workshop_rolling-update/src/exercises/Solutions.hs [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+
+
+
+-- Given
+-- Use `fmap` for IO to implement this function!
+-- Useful helpers:
+--   - readFile :: String -> IO String
+--   - length :: String -> Int
+getFileLength :: FilePath -> IO Int
+getFileLength name = ((fmap length) . readFile) name
```
Comment (quchen) [inline beginner-workshop_rolling-update/src/exercises/Solutions.hs]: Redundant parenthesis

File: beginner-workshop_rolling-update/src/exercises/Solutions.hs [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+everyOther :: [a] -> [a]
+everyOther (x:y:xs) = [x] ++ everyOther xs
+everyOther _ = []
+
+-- Split a list into halves of equal length (±1).
+-- Easy using `length`, tricky so it works even on infinite lists.
+splitHalf :: [a] -> ([a], [a])
+splitHalf (x:y:xs) = let (xs', ys') = splitHalf xs
+  in (([x]++xs'), ([y]++ys'))
+splitHalf _ = ([], [])
```
Comment (quchen) [inline beginner-workshop_rolling-update/src/exercises/Solutions.hs]: Correct solution, the question is a bit fuzzy. 

What I meant was a function splitting the list "in the middle", splitHalf [1,2,3,4]=([1,2],[3,4])

File: beginner-workshop_rolling-update/src/exercises/Solutions.hs [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+          guessingGame number
+        Just guessInt -> 
+          case compare number guessInt of
+            LT -> do
+              putStrLn "You guess is too high"
+              guessingGame number
+            GT -> do
+              putStrLn "Your guess is too low"
+              guessingGame number
+            EQ -> putStrLn "You cheated!"
```
Comment (quchen) [inline beginner-workshop_rolling-update/src/exercises/Solutions.hs]: Haha :-P

**Comment by fdw** [comment]
Thanks!

I changed the small stuff, and I tried to solve the new task. However, I only found the obvious solution using `length` :(
Could you give me a hint how to solve it recursively (as I see no other solution that works on infinite lists)? Or did I miss something else?

**Comment by quchen** [comment]
Find out what 'zipWith const' does.
Use it to implement 'dropLast' for possibly infinite lists. (drop the last element, like take (length xs - 1) xs for finite lists) 
Use it to implement 'dropLast n'  to drop the last n elements of a possibly infinite list. 
Now revisit the problem with your newfound knowledge about 'zipWith last'. 
It's a pretty difficult exercise, just ask me online of you need more help. The nice thing is that it's at least an interesting thing to see that we don't need to know the length for lots of applications.

**Comment by fdw** [comment]
So, with your help I finally found a solution :)

**Pull request merged**